### ISSUE TITLE
Proposal to decrease storage space requirements in data model

### ISSUE THREAD

**Issue by thesunny** [ianstormtaylor/slate#3311]
#### Do you want to request a _feature_ or report a _bug_?

Proposal to change data format to balance storage space against readability

This issue re-opened from here as suggested by Ian https://github.com/ianstormtaylor/slate/issues/3276

#### What's the current behavior?

The current data model is verbose (i.e. property keys are described in full).

Given that the data model is pure JSON since 0.50, it will be more common for developers to save Slate data into the database without an encode/decode step. For this reason, there is some benefit in considering balancing readability against storage requirements for Slate data.

#### What's the expected behavior?

Here are three proposals that trade some readability for improved storage space along with their space savings on the rich text demo data. The test data is probably optimistic as there is a lot of formatting. On the other hand, tables and other nested structure may optimize even better.

- convert `text` property to `x` (6.7% savings on rich text demo)
- convert `children` to `ch` (4.4% savings on rich text demo)
- best practices to use tag names versus expanded names for type values in demos and tutorials (6.5% savings)

Combining all changes results in an 18% savings in storage space when the data structure is minified.

I suggest to consider each change individually rather than as a group. For example, the best practice of using tag names nets us a decent space saving without reducing readability because developers are already familiar with tag names like `p`, `h1`, etc.

Here are examples of how this would look.

```js
// original rich text format
const x = [
  {
    type: "paragraph",
    children: [
      { text: "This is editable " },
      { text: "rich", bold: true },
      { text: " text, " },
      { text: "much", italic: true },
      { text: " better than a " },
      { text: "<textarea>", code: true },
      { text: "!" },
    ],
  },
  {
    type: "paragraph",
    children: [
      { text: "Since it's rich text, you can do things like turn a selection of text " },
      { text: "bold", bold: true },
      { text: ", or add a semantically rendered block quote in the middle of the page, like this:" },
    ],
  },
  {
    type: "block-quote",
    children: [{ text: "A wise quote." }],
  },
  {
    type: "paragraph",
    children: [{ text: "Try it out for yourself!" }],
  },
]

// replace `text` with `x`
const x = [
  {
    type: "paragraph",
    children: [
      { x: "This is editable " },
      { x: "rich", bold: true },
      { x: " text, " },
      { x: "much", italic: true },
      { x: " better than a " },
      { x: "<textarea>", code: true },
      { x: "!" },
    ],
  },
  {
    type: "paragraph",
    children: [
      { x: "Since it's rich text, you can do things like turn a selection of text " },
      { x: "bold", bold: true },
      { x: ", or add a semantically rendered block quote in the middle of the page, like this:" },
    ],
  },
  {
    type: "block-quote",
    children: [{ x: "A wise quote." }],
  },
  {
    type: "paragraph",
    children: [{ x: "Try it out for yourself!" }],
  },
]

// also replace `children` with `ch`
const x = [
  {
    type: "paragraph",
    ch: [
      { x: "This is editable " },
      { x: "rich", bold: true },
      { x: " text, " },
      { x: "much", italic: true },
      { x: " better than a " },
      { x: "<textarea>", code: true },
      { x: "!" },
    ],
  },
  {
    type: "paragraph",
    ch: [
      { x: "Since it's rich text, you can do things like turn a selection of text " },
      { x: "bold", bold: true },
      { x: ", or add a semantically rendered block quote in the middle of the page, like this:" },
    ],
  },
  {
    type: "block-quote",
    ch: [{ x: "A wise quote." }],
  },
  {
    type: "paragraph",
    ch: [{ x: "Try it out for yourself!" }],
  },
]

// Also best practices to favour tag names to longer spelled out name
const x = [
  {
    type: "p",
    ch: [
      { x: "This is editable " },
      { x: "rich", b: true },
      { x: " text, " },
      { x: "much", i: true },
      { x: " better than a " },
      { x: "<textarea>", code: true },
      { x: "!" },
    ],
  },
  {
    type: "p",
    ch: [
      { x: "Since it's rich text, you can do things like turn a selection of text " },
      { x: "bold", b: true },
      { x: ", or add a semantically rendered block quote in the middle of the page, like this:" },
    ],
  },
  {
    type: "block-quote",
    ch: [{ x: "A wise quote." }],
  },
  {
    type: "p",
    ch: [{ x: "Try it out for yourself!" }],
  },
]
```

**Comment by ianstormtaylor** [comment]
What are the savings once GZIP is factored in?

**Comment by thesunny** [comment]
I forgot to mention that for the comparison above, I minified the data before taking measurements.

In answer to your question, with minification then gzip, the demo files turned in sizes of:

- 288 bytes for original rich text data
- 273 bytes for all the changes applied

Savings of 5.2% down from 18%

For clarity, the proposal above was more for database storage than for data transfer.

To bring more clarity to that question, did some poking around on compression in JSON fields for databases and found this:

- Postgres doesn't appear to use any compression on its `jsonb` format but there is a [zson plugin](https://github.com/postgrespro/zson)
- MongoDB uses [Snappy](https://github.com/google/snappy) compression which increases speed at the cost of amount compressed. They says this: "For instance, compared to the fastest mode of zlib, Snappy is an order of magnitude faster for most inputs, but the resulting compressed files are anywhere from 20% to 100% bigger." 
- MySQL doesn't seem to have compression on by default but it can be [enabled by table](https://dev.mysql.com/doc/refman/5.6/en/innodb-compression.html)

**Comment by ianstormtaylor** [comment]
I don't think Postgres is uncompressed for JSON(B) columns by default, last I checked.

I have a feeling this is going to depend too much on the actual domain-specific needs of each user to do anything about in core. People will have varying document lengths, average words-per-block, nesting of element nodes, ratio of custom properties, etc. 

Reducing usability/readability for potentially single-digit-percentage gains doesn't seem worth it? Especially because use cases that have enough scale to need worry about data size will likely need to have their own serialization steps before putting things into the database anyways.

If people want to add real-world data from compressing documents they've seen in the wild for their use case though that might make the argument stronger. This is a case where we shouldn't do anything without good quality data to back it up.

**Comment by thesunny** [comment]
To add some context, I've been using `x` for `text` fields and tag names instead of full names for over a year in my storage format that I use for Slate.

I have found them to be readable and easier to generate test data because the data format is very concise.

```js
[{type: 'p', children: [{x: 'This is text'}]}
```

I do not use `ch` for Slate currently but recently, in another library for handling trees, I used `ch` instead of `children` and have found that although it is a little less readable it was not hard to adjust to. I experimented with other common properties like `type` to `tp` but found that they were confusing.

I realized the reason for this is that `{x: 'some text here'}` gives context because the text itself provides a hint. Similarly, but less so, `{ch: [{...}, {...}, {...}]]` provides a hint to the meaning of ch because it's value is an array of objects (i.e. children). But I found `{tp: 'p'}` harder to intuit which is why I haven't added a proposal for renaming other properties.

Overall, I'm more positive towards the `x` and the tag name practices and am on the fence about `ch`. That said, I don't think it's that important but since we are making a lot of breaking changes, it felt like this was the best time to bring it up for consideration.
